### PR TITLE
support Nacos cluster

### PR DESCRIPTION
**Pull request by IanCao**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Comment by wu-sheng** [comment]
Please check Nacos license issue first, then we could talk about this PR.

**Comment by IanCao** [comment]
> Please check Nacos license issue first, then we could talk about this PR.

Nacos is under  Apache License  Version 2.0

**Comment by wu-sheng** [comment]
@IanCao That is not enough, read https://github.com/apache/skywalking/pull/2422 , we need to scan all dependencies because you are contributing features at server side, which means we will re-distribute their jar(s) in our binary.  In agent plugin, we just use them to compile.

**Comment by IanCao** [comment]
ok, let me check

**Comment by IanCao** [comment]
image

```


Group ID: io.prometheus 
Artifact ID: simpleclient
Version: 0.5.0
Apache License 2.0.

Group ID: com.fasterxml.jackson.core 
Artifact ID: jackson-databind 
Version: 2.9.8
Apache License 2.0.


Group ID: com.fasterxml.jackson.core 
Artifact ID: jackson-annotations 
Version: 2.9.0
Apache License 2.0.

Group ID: com.fasterxml.jackson.core 
Artifact ID: jackson-core 
Version: 2.9.8
Apache License 2.0.

Group ID: commons-codec 
Artifact ID: commons-codec 
Version: 1.11
Apache License 2.0.

Group ID: com.alibaba.nacos 
Artifact ID: nacos-api 
Version: 1.0.0
Apache License 2.0.

Group ID: com.alibaba 
Artifact ID: fastjson 
Version: 1.2.47
Apache License 2.0.

Group ID: com.alibaba.nacos 
Artifact ID: nacos-common 
Version: 1.0.0
Apache License 2.0.

Group ID: org.apache.commons 
Artifact ID: commons-lang3 
Version: 3.4
Apache License 2.0.

Group ID: commons-io 
Artifact ID: commons-io 
Version: 2.6
Apache License 2.0.
```

all dependencies is under apache license 2.0

**Comment by wu-sheng** [comment]
There is something missing in the screenshot, could you provide the difference between before and after the nacos dependency? Which libs version update? Which libs are added?

**Comment by IanCao** [comment]
> re added?
the left line point to com.google.guava . 
and i'll show the difference

**Comment by IanCao** [comment]
> There is something missing in the screenshot, could you provide the difference between before and after the nacos dependency? Which libs version update? Which libs are added?

the check result： five jars added, no jar's version update.
the five jar list:
1. cluster-nacos-plugin-6.2.0-SNAPSHOT.jar
2. fastjson-1.2.47.jar
3. nacos-api-1.0.0.jar
4. nacos-client-1.0.0.jar
5. nacos-common-1.0.0.jar

**Comment by IanCao** [comment]
i have updated the license, plz review

File: apm-dist/release-docs/NOTICE [hunk 1]
```diff
@@ -831,3 +831,61 @@ from and not be held liable to the user for any such damages as noted
...
+
+Also, please refer to each LICENSE.<component>.txt file, which is located in
+the 'license' directory of the distribution file, for the license terms of the
+components that this product depends on.
+
+------
+
+===========================================================================
+fastjson-1.2.47 Notice
+===========================================================================
```
Comment (wu-sheng) [inline apm-dist/release-docs/NOTICE]: fastjson notice should be removed. It doesn't include more things than its license file.

File: apm-dist/release-docs/licenses/LICENSE-nacos.txt [hunk 2]
```diff
@@ -0,0 +1,201 @@
+  Apache License
```
Comment (wu-sheng) [inline apm-dist/release-docs/licenses/LICENSE-nacos.txt]: This looks like a standard Apache license, which could not include the text. If I am right, please remove this file.

File: docs/en/setup/backend/backend-cluster.md [hunk 3]
```diff
@@ -73,6 +73,18 @@ cluster:
     hostPort: ${SW_CLUSTER_CONSUL_HOST_PORT:localhost:8500}
 ```
 
+## Nacos
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-cluster.md]: Why do you add document inside `Consul` paragraph?

**Comment by wu-sheng** [comment]
@peng-yongsheng Please recheck the document/LICENSE first, the new Nacos version with the exclusions should be good.

File: apm-dist/release-docs/licenses/LICENSE-nacos.txt (hunk 2 above)
Comment (IanCao) [reply apm-dist/release-docs/licenses/LICENSE-nacos.txt]: i just found the okhttp license is same as nacos.
and okhttp license appear in apm-dist/release-docs/licenses/
so i put nacos license into there

File: docs/en/setup/backend/backend-cluster.md (hunk 3 above)
Comment (IanCao) [reply docs/en/setup/backend/backend-cluster.md]: consul contains hostPort and serviceName, nacos also contains hostPort and serviceName,so i just copy it and change the cluster Name and default port

File: apm-dist/release-docs/NOTICE (hunk 1 above)
Comment (IanCao) [reply apm-dist/release-docs/NOTICE]: ok

File: docs/en/setup/backend/backend-cluster.md (hunk 3 above)
Comment (wu-sheng) [reply docs/en/setup/backend/backend-cluster.md]: I am talking about the document, the following document is part of Consul document, if your Nacos document needs this too, please copy it in your paragraph. 
> Same as Zookeeper coordinator, in some cases, oap default gRPC host and port in core are not suitable for internal communication among the oap nodes. The following setting are provided to set the hot and port manually, based on your own LAN env.
internalComHost, the host registered and other oap node use this to communicate with current node.
internalComPort, the port registered and other oap node use this to communicate with current node.

Right now, it looks like Consule doesn't support his, see at https://github.com/apache/skywalking/blob/dc39825a74ad31df0c1ab2736d3fed4a2f31382d/docs/en/setup/backend/backend-cluster.md#consul

File: apm-dist/release-docs/licenses/LICENSE-nacos.txt (hunk 2 above)
Comment (wu-sheng) [reply apm-dist/release-docs/licenses/LICENSE-nacos.txt]: This is a suggestion, we don't need to copy all LICENSE when it is standard Apache 2.0.

File: docs/en/setup/backend/backend-cluster.md (hunk 3 above)
Comment (wu-sheng) [reply docs/en/setup/backend/backend-cluster.md]: `internalComHost` and `internalComPort` things are not supported in Nacos. Suggest to add.

Comment (IanCao) [reply docs/en/setup/backend/backend-cluster.md]: oo..its my fault.
I have not noticed that.
i'll fix

**Comment by peng-yongsheng** [comment]
Hi @IanCao, can you help me to correct some mistakes about not clear dependency management. I checked the dependency tree. And I found those third-party libraries are not managed by the OAP module's dependency manager, etc.. jackson-core, jackson-annotations, jackson-databind, commons-codec, commons-lang3.
This is not caused by you. Those libraries are not managed before. ^_^

**Comment by wu-sheng** [comment]
@peng-yongsheng could we change those through different pull request? Look like irrelevant.

**Comment by peng-yongsheng** [comment]
image

[Image source](https://mvnrepository.com/artifact/com.alibaba.nacos/nacos-client/1.0.0)

From this compile dependencies list, his exclude list is right and the third-party library's licenses are no problem. @wu-sheng

**Comment by peng-yongsheng** [comment]
> @peng-yongsheng could we change those through different pull request? Look like irrelevant.

Those third-party libraries are using by gPRC and zookepper, but with different version. I'm not sure the OAP server is work well after the version changed.

**Comment by peng-yongsheng** [comment]
> @peng-yongsheng could we change those through different pull request? Look like irrelevant.

@wu-sheng I think keeping those third-party libraries version is the best solution. He can only test the nacos client without test the whole OAP server. So, he can find the version number from current master source then put them into the dependency management.

**Comment by wu-sheng** [comment]
@peng-yongsheng I am just suggesting, we could try to work on another pull request to make sure the dependency clear for current master, then we update this pr.

**Comment by IanCao** [comment]
> Hi @IanCao, can you help me to correct some mistakes about not clear dependency management. I checked the dependency tree. And I found those third-party libraries are not managed by the OAP module's dependency manager, etc.. jackson-core, jackson-annotations, jackson-databind, commons-codec, commons-lang3.
> This is not caused by you. Those libraries are not managed before. ^_^

ok, i'll make the dependency management more clear and open a new pr to fix

**Pull request merged**